packages/app-project/src/components/AuthModal/shared/components/withFormik/withFormik.js

```diff
@@ -0,0 +1,64 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import { Formik } from 'formik'
+
+function withFormik (WrappedComponent) {
+  class HOC extends React.Component {
+    constructor () {
+      super()
+
+      this.handleBlurWithCallback = this.handleBlurWithCallback.bind(this)
+      this.handleChangeWithCallback = this.handleChangeWithCallback.bind(this)
+    }
+
+    handleBlurWithCallback (event, handleBlur, formikProps) {
+      handleBlur(event)
+      this.props.onBlur(event, formikProps)
+    }
+
+    handleChangeWithCallback (event, handleChange, formikProps) {
+      handleChange(event)
+      this.props.onChange(event, formikProps)
+    }
+
+    render () {
+      const { initialValues, onSubmit } = this.props
+      return (
+        <Formik
+          initialValues={initialValues}
+          onSubmit={onSubmit}
+          {...this.props}
+        >
+          {(innerProps) => {
+            const {
+              handleBlur,
+              handleChange
+            } = innerProps
+
+            return (
+              <WrappedComponent
+                handleBlur={(event) => this.handleBlurWithCallback(event, handleBlur, innerProps)}
+                handleChange={(event) => this.handleChangeWithCallback(event, handleChange, innerProps)}
+                {...innerProps}
```

Comment:
Will this overwrite `handleChange` and `handleBlur` with `innerProps.handleChange` and `innerProps.handleBlur`?

Reply:
I'll write up a readme about what this is doing. This is adding support for adding custom callbacks to these two events if we want to do our own async validations on these events. Formik doesn't directly support this. I can see how it's confusing with the naming scheme going on.

Reply:
One solution would be to remove `handleBlur` and `handleChange` from the spread props, so that's there no chance of accidental duplication.

Reply:
Sorry, the readme has left me more confused. The function now passes new event handlers down in addition to the original handlers passed down in `...innerProps`. So `WrappedComponent` has a `handleChange` prop and an `onChangeEvent` prop and so on for blur, reset and submit. Is one handler supposed to handle async validation and the other synchronous validation. I'm finding this code very hard to understand.

Reply:
`handleChange` is the original Formik handler that is required to be used with the inputs on change event. Their internal API is handling the form state and updating the input during the on change event. I've added support for the `onChange` prop since Formik doesn't support this, so we can support things like async validation on the on change event.

Similarly I did this with on blur too because Formik itself doesn't support customization of the event handling. They do support customization of the submit and reset event, but I've explicitly set them to `onResetEvent` and `onSubmitEvent` to make the prop names consistent for the consuming form. 

I'll try another pass at the naming of things and the readme since it's still not clear.

Reply:
OK. What's confusing me most at the moment is that in the React dev tools I can see `handleChange` in the wrapped Form's props (from `{...innerProps}`) but the code for `Form.js` doesn't use that prop, unless I've missed it.

I _think_ it's fine to name it `handleChange`, as long as you don't also use `{...innerProps}` to pass `handleChange` down to the wrapped form.

Reply:
Yeah, I need to fix how I'm destructuring and passing the props.